File: java/src/test/java/io/cucumber/cucumberexpressions/CucumberExpressionTokenizerTest.java
package io.cucumber.cucumberexpressions;

import io.cucumber.cucumberexpressions.Ast.Token;
import io.cucumber.cucumberexpressions.Ast.Token.Type;
import org.junit.jupiter.api.extension.ParameterContext;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.converter.ArgumentConversionException;
import org.junit.jupiter.params.converter.ArgumentConverter;
import org.junit.jupiter.params.converter.ConvertWith;
import org.junit.jupiter.params.provider.MethodSource;
import org.yaml.snakeyaml.Yaml;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

import static java.nio.file.Files.newDirectoryStream;
import static java.nio.file.Files.newInputStream;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.junit.jupiter.api.Assertions.assertThrows;

class CucumberExpressionTokenizerTest {

    private final CucumberExpressionTokenizer tokenizer = new CucumberExpressionTokenizer();

    private static List<Path> acceptance_tests_pass() throws IOException {
        List<Path> paths = new ArrayList<>();
        newDirectoryStream(Paths.get("..", "testdata", "cucumber-expression", "tokenizer")).forEach(paths::add);
        paths.sort(Comparator.naturalOrder());
        return paths;
    }

    @ParameterizedTest
    @MethodSource
    void acceptance_tests_pass(@ConvertWith(Converter.class) Expectation expectation) {
        if (expectation.exception == null) {
            List<Token> tokens = tokenizer.tokenize(expectation.expression);
            List<Token> expectedTokens = expectation.expected_tokens
                    .stream()
                    .map(YamlableToken::toToken)
                    .collect(Collectors.toList());
            assertThat(tokens, is(expectedTokens));
        } else {
            CucumberExpressionException exception = assertThrows(
                    CucumberExpressionException.class,
                    () -> tokenizer.tokenize(expectation.expression));
            assertThat(exception.getMessage(), is(expectation.exception));
        }
    }

    static class Expectation {
        public String expression;
        public List<YamlableToken> expected_tokens;
        public String exception;
    }

    static class Converter implements ArgumentConverter {
        Yaml yaml = new Yaml();

        @Override
        public Expectation convert(Object source, ParameterContext context) throws ArgumentConversionException {
            try {
                Path path = (Path) source;
                InputStream inputStream = newInputStream(path);
                return yaml.loadAs(inputStream, Expectation.class);
            } catch (IOException e) {
                throw new ArgumentConversionException("Could not load " + source, e);
            }
        }
    }

    static class YamlableToken {
        public String text;
        public Type type;
        public int start;
        public int end;

        public Token toToken() {
            return new Token(text, type, start, end);
        }
    }
}

File: elixir/test/cucumber/cucumber_expressions/cucumber_expression_tokenizer_test.exs
defmodule Cucumber.CucumberExpressions.CucumberExpressionTokenizerTest do
  use ExUnit.Case, async: true

  alias Cucumber.CucumberExpressions.{Testdata, Token, CucumberExpressionTokenizer}

  for {name, fixture} <- Testdata.load("cucumber-expression/tokenizer") do
    @fixture fixture
    if Map.has_key?(fixture, "exception") do
      test "rejects #{name}" do
        %{"expression" => expression, "exception" => expected} = @fixture
        assert {:error, error} = CucumberExpressionTokenizer.tokenize(expression)
        assert Exception.message(error) == expected
      end
    else
      test "tokenizes #{name}" do
        %{"expression" => expression, "expected_tokens" => expected} = @fixture
        assert {:ok, tokens} = CucumberExpressionTokenizer.tokenize(expression)
        assert Enum.map(tokens, &token_to_map/1) == expected
      end
    end
  end

  defp token_to_map(%Token{type: type, text: text, start: start, end: end_}) do
    %{"type" => Token.type_string(type), "start" => start, "end" => end_, "text" => text}
  end
end
